server/src/main/java/org/apache/cassandra/sidecar/job/JobTracker.java

```diff
@@ -0,0 +1,254 @@
+        try
+        {
+            return super.put(key, value);
+        }
+        finally
+        {
+            lock.writeLock().unlock();
+        }
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public Job putIfAbsent(final UUID key, final Job value)
+    {
+        lock.writeLock().lock();
+        try
+        {
+            return super.putIfAbsent(key, value);
+        }
+        finally
+        {
+            lock.writeLock().unlock();
+        }
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public void putAll(final Map<? extends UUID, ? extends Job> m)
+    {
+        lock.writeLock().lock();
+        try
+        {
+            super.putAll(m);
+        }
+        finally
+        {
+            lock.writeLock().unlock();
+        }
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public Job remove(final Object key)
+    {
+        lock.writeLock().lock();
+        try
+        {
+            return super.remove(key);
+        }
+        finally
+        {
+            lock.writeLock().unlock();
+        }
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public void clear()
+    {
+        lock.writeLock().lock();
+        try
+        {
+            super.clear();
+        }
+        finally
+        {
+            lock.writeLock().unlock();
+        }
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public Job get(final Object key)
+    {
+        lock.readLock().lock();
+        try
+        {
+            return super.get(key);
+        }
+        finally
+        {
+            lock.readLock().unlock();
+        }
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public boolean containsKey(final Object key)
+    {
+        lock.readLock().lock();
+        try
+        {
+            return super.containsKey(key);
+        }
+        finally
+        {
+            lock.readLock().unlock();
+        }
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public int size()
+    {
+        lock.readLock().lock();
+        try
+        {
+            return super.size();
+        }
+        finally
+        {
+            lock.readLock().unlock();
+        }
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public boolean isEmpty()
+    {
+        lock.readLock().lock();
+        try
+        {
+            return super.isEmpty();
+        }
+        finally
+        {
+            lock.readLock().unlock();
+        }
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public Object clone()
+    {
+        lock.readLock().lock();
+        try
+        {
+            return super.clone();
+        }
+        finally
+        {
+            lock.readLock().unlock();
+        }
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    protected boolean removeEldestEntry(final Map.Entry<UUID, Job> eldest)
+    {
+        if (size() <= capacity)
```

Comment:
Let's say capacity is 5 (just assuming for the discussion), all 5 jobs are ongoing, if a new job comes up do we force delete the eldest ongoing job from tracker? If so, how can a user track that deleted eldest job?
Instead of having capacity, if we are just worried about too many ongoing jobs and jobtracker taking up too much memory, can we have a cap on ongoing jobs and reject incoming requests if jobtracker reached the maximum limit and ask user to wait to finish an ongoing job to submit a new one?

Reply:
So, the intention behind the tracker is only to maintain the state for ongoing jobs without growing unbounded, but not to throttle/limit the no. of jobs that can be posted.

Since jobs are typically short-lived, the entry's age based eviction seems reasonable as a capping mechanism. 

> how can a user track that deleted eldest job
Ideally, we should size this optimally to minimize evictions of ongoing jobs.
If a job that is inflight on the C* node gets evicted from the cache, the subsequent polling will return a 404. 

The expected behavior for the client at this point is to retry the command (eg. decommission), here the downstream check will determine that there is an ongoing operation and return a 409 Conflict, and start tracking it with a new UUID (which is part of the response).

Reply:
I am still not convinced as this doesn't feel like smooth user experience. Returning 404 after Sidecar restart sounds fine. But without any process restart, kicking out a ongoing job from the map and user seeing 404 doesn't feel like great user experience. It will confuse the user who just submitted the job, polling for status and seeing 404.